File: chapter16_2_java/app/src/main/java/com/example/chapter16_2_java/MainActivity.java
package com.example.chapter16_2_java;

import android.content.ContentValues;
import android.net.Uri;
import android.os.Bundle;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ListView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;

import java.util.ArrayList;

public class MainActivity extends AppCompatActivity {
    private ArrayList<String> items = new ArrayList<>();
    private ArrayAdapter<String> adapter;
    private final Uri uri = Uri.parse("content://com.example.lab16");

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        enableEdgeToEdge();
        setContentView(R.layout.activity_main);
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main), (v, insets) -> {
            Insets systemBarsInsets = insets.getInsets(WindowInsetsCompat.Type.systemBars());
            v.setPadding(systemBarsInsets.left, systemBarsInsets.top, systemBarsInsets.right, systemBarsInsets.bottom);
            return insets;
        });

        adapter = new ArrayAdapter<>(this, android.R.layout.simple_list_item_1, items);
        ((ListView) findViewById(R.id.listView)).setAdapter(adapter);
        setListener();
    }

    private void setListener() {
        EditText edBook = findViewById(R.id.edBook);
        EditText edPrice = findViewById(R.id.edPrice);

        findViewById(R.id.btnInsert).setOnClickListener(v -> {
            String name = edBook.getText().toString();
            String price = edPrice.getText().toString();
            if (name.isEmpty() || price.isEmpty()) {
                showToast("欄位請勿留空");
            } else {
                ContentValues values = new ContentValues();
                values.put("book", name);
                values.put("price", price);
                Uri contentUri = getContentResolver().insert(uri, values);
                if (contentUri != null) {
                    showToast("新增:" + name + ",價格:" + price);
                    cleanEditText();
                } else {
                    showToast("新增失敗");
                }
            }
        });

        findViewById(R.id.btnUpdate).setOnClickListener(v -> {
            String name = edBook.getText().toString();
            String price = edPrice.getText().toString();
            if (name.isEmpty() || price.isEmpty()) {
                showToast("欄位請勿留空");
            } else {
                ContentValues values = new ContentValues();
                values.put("price", price);
                int count = getContentResolver().update(uri, values, name, null);
                if (count > 0) {
                    showToast("更新:" + name + ",價格:" + price);
                    cleanEditText();
                } else {
                    showToast("更新失敗");
                }
            }
        });

        findViewById(R.id.btnDelete).setOnClickListener(v -> {
            String name = edBook.getText().toString();
            if (name.isEmpty()) {
                showToast("書名請勿留空");
            } else {
                int count = getContentResolver().delete(uri, name, null);
                if (count > 0) {
                    showToast("刪除:" + name);
                    cleanEditText();
                } else {
                    showToast("刪除失敗");
                }
            }
        });

        findViewById(R.id.btnQuery).setOnClickListener(v -> {
            String name = edBook.getText().toString();
            String selection = name.isEmpty() ? null : name;
            try (android.database.Cursor c = getContentResolver().query(uri, null, selection, null, null)) {
                if (c == null) return;
                c.moveToFirst();
                items.clear();
                showToast("共有" + c.getCount() + "筆資料");
                for (int i = 0; i < c.getCount(); i++) {
                    items.add("書名:" + c.getString(0) + "\t\t\t\t價格:" + c.getInt(1));
                    c.moveToNext();
                }
                adapter.notifyDataSetChanged();
            }
        });
    }

    private void showToast(String text) {
        Toast.makeText(this, text, Toast.LENGTH_LONG).show();
    }

    private void cleanEditText() {
        ((EditText) findViewById(R.id.edBook)).setText("");
        ((EditText) findViewById(R.id.edPrice)).setText("");
    }

    private void enableEdgeToEdge() {

    }
}

File: Kotlin HW1/app/src/main/java/com/example/myapplication/MainActivity.kt
package com.example.myapplication

import android.os.Bundle
import android.widget.Button
import android.widget.EditText
import android.widget.RadioGroup
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat

class MainActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main)) { v, insets ->
            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
            insets
        }

        val edName = findViewById<EditText>(R.id.edName)
        val tvText = findViewById<TextView>(R.id.tvText)
        val radioGroup = findViewById<RadioGroup>(R.id.radioGroup)
        val btnMora = findViewById<Button>(R.id.btnMora)
        val tvName = findViewById<TextView>(R.id.tvName)
        val tvWinner = findViewById<TextView>(R.id.tvWinner)
        val tvMyMora = findViewById<TextView>(R.id.tvMyMora)
        val tvTargetMora = findViewById<TextView>(R.id.tvTargetMora)

        btnMora.setOnClickListener {
            if (edName.text.isEmpty()) {
                tvText.text = "請輸入玩家姓名"
                return@setOnClickListener
            }

            val playerName = edName.text.toString()
            val targetMora = (0..2).random()
            val myMora = when (radioGroup.checkedRadioButtonId) {
                R.id.btnScissor -> 0
                R.id.btnStone -> 1
                else -> 2
            }

            tvName.text = "名字\n$playerName"
            tvMyMora.text = "我方出拳\n${getMoraString(myMora)}"
            tvTargetMora.text = "電腦出拳\n${getMoraString(targetMora)}"

            val winner = when {
                myMora == targetMora -> "平手"
                (myMora == 0 && targetMora == 2) ||
                        (myMora == 1 && targetMora == 0) ||
                        (myMora == 2 && targetMora == 1) -> playerName
                else -> "電腦"
            }

            tvWinner.text = "勝利者\n$winner"
            tvText.text = if (winner == "平手") "平局，請再試一次！" else if (winner == playerName) "恭喜你獲勝了！！！" else "可惜，電腦獲勝了！"
        }
    }

    private fun getMoraString(mora: Int) = when (mora) {
        0 -> "剪刀"
        1 -> "石頭"
        else -> "布"
    }
}